### DISCUSSION TITLE
Page metrics

### DISCUSSION THREAD

**Discussion by inancgumus** [grafana/xk6-browser#190] [Ideas]
One of our users wanted to measure the response time of a page (here: #189) using `sleep` and `Date.now`. This resulted in zero difference (`t1 - t0` == 0):

```javascript
var src = page.goto('http://...', { waitUntil: ‘networkidle’ });
sleep(0.3);

var t0 = Date.now();
page.waitForLoadState(‘networkidle’);
var t1 = Date.now();
console.log((t1-t0));
```

As we discovered before, [here](https://github.com/grafana/xk6-browser/issues/125#issuecomment-976678635):
* **Don't Sleep:** The issue doesn't occur if we don't sleep in the test code. I determined this yesterday as well, however, it should be the correct behavior for test scripts the follow which is "never sleep"! [Note that _Playwright also recommends the same._](https://playwright.dev/python/docs/intro/#known-issues)

@imiric suggested:
> In the future, we might want to expose page metrics similar to Puppeteer's, so that you can get this internally instead of relying on custom time measurements, but we haven't discussed or planned this feature yet.
>
> It seems this is [not implemented in Playwright](https://github.com/microsoft/playwright/issues/590) either, which is probably why we missed it, but it looks like a good idea to have for these use cases.

So, implementing `page.metrics` might be a good idea. This discussion is the place for doing it :)

**Comment by robingustafsson** [comment]
We definitely want to expose page load time metrics (and probably some/all of the ones exposed by Puppeteer). Using sleeps for this will never work reliably, the tool/browser needs to provide these measurements.

A seperate topic really, but we also want to support [Lighthouse](https://developers.google.com/web/tools/lighthouse/) type analysis output eventually as well, see https://github.com/GoogleChrome/lighthouse/blob/HEAD/docs/readme.md#using-programmatically for example on how to invoke Lighthouse. And [here](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/gather/gatherers) for all the "gatherers" of data (Lighthouse interfaces with the browser over CDP to collect this data).

**Comment by inancgumus** [comment]
Hmm, Lighthouse looks interesting. Thanks for sharing it. I looked into it a little bit, and it seems like it's also some sort of a CDP protocol listener (roughly similar to xk6-browser), but for a different purpose than ours, for producing audit reports. Since we're already listening for CDP events, so, yeah, we can build an audit tool like Lighthouse in Go.

Maybe we can produce trace output for users who want to process it through Lighthouse as well, [like this](https://github.com/GoogleChrome/lighthouse/blob/HEAD/docs/readme.md#lighthouse-as-trace-processor)—afterwards, of course :)
